docs: release note for Metals 0.11.6 (#3957)
scalameta/metals: website/blog/2022-05-27-aluminium.md

MUnit allows to use [helper functions](https://scalameta.org/munit/docs/tests.html#declare-tests-inside-a-helper-function) when declaring tests. Very often those helper methods are extracted to some parent classes which are extended by many test suites.
Now, Metals can find usages of those helper methods and display them in Test Explorer.

This feature is available for Bloop and SBT 1.7.0-M2 or later.

![MUnit-helper-methods](https://imgur.com/GGRDpXA.gif)

## Support Cats Effect stacktraces in stacktrace analyzer

Cats Effect offers [asynchronous stack tracing](https://typelevel.org/cats-effect/docs/tracing#asynchronous-stack-tracing) which augments exceptions with additional information.  
Now, Stacktrace Analyzer is able to recognize CE's stacktraces and provide link to location in code. Say no to tedious debugging when you have only stacktrace from the logs!

![cats-effect-stacktraces](https://imgur.com/5fMvcYd.gif)

## Improve implement all completion and code action
Previously, when we invoked the "implement all members" completion and code action, Metals used `x$0` for Java parameter names. Now, Metals fills these with the correct parameter names.

>  
> 

Reviewer: ```suggestion
## Improve rewrite to braces/parenthesis code action

Sometimes it wasn't clear which code would be affected by [rewrite to braces/parenthesis](https://scalameta.org/metals/blog/2021/09/06/tungsten#replace--with--in-functions-and-vice-versa) code action.
Now, code action's description contains name of the function/method which will be affected by executing action.

![rewrite-braces-parens](https://imgur.com/SkHolsJ.gif)
Theme: [One Dark Pro](https://marketplace.visualstudio.com/items?itemName=zhuangtongfa.Material-theme)

```

Reviewer: I think it's handy to include theme name for each GIF that was included in release notes

Author: Thank you for adding words to a release note!

> I think it's handy to include the theme name for each GIF that was included in release notes

I personally think I don't need to add a theme name to release notes because most of the readers of this note are interested in what's new in the latest Metals. Color theme is sometimes interesting, but they are not related to the Metals.

---
author: Rikito Taniguchi
title: Metals v0.11.6 - Aluminium
authorURL: https://twitter.com/tanishiking25
authorImageURL: https://github.com/tanishiking.png
---

We're happy to announce the release of Metals v0.11.6, which 

> - reduce memory usage in large projects

Reviewer 2: ```suggestion
- reduces memory usage in large projects
```

> - add override completion for Scala3

Reviewer 2: ```suggestion
- adds override completion for Scala3
```

> - improve scaladoc support both in Scala2 and Scala3

Reviewer 2: ```suggestion
- improves scaladoc support in both Scala2 and Scala3
```

> - provide better UX in test explorer

Reviewer 2: ```suggestion
- improves the UX in the test explorer
```

- support for Scala 3.1.3-RC3 and RC4

<table>
<tbody>
  <tr>
    <td>Commits since last release</td>
    <td align="center">148</td>
  </tr>
  <tr>
    <td>Merged PRs</td>
    <td align="center">48</td>
  </tr>
    <tr>
    <td>Contributors</td>
    <td align="center">12</td>
  </tr>
  <tr>
    <td>New features</td>
    <td align="center">7</td>
  </tr>
</tbody>
</table>

Metals is a language server for Scala that works with VS Code, Vim, Emacs and
Sublime Text. Metals is developed at the
[Scala Center](https://scala.epfl.ch/) and [VirtusLab](https://virtuslab.com)
with the help from [Lunatech](https://lunatech.com) along with contributors from
the community.

## TL;DR

Check out [https://scalameta.org/metals/](https://scalameta.org/metals/), and
give Metals a try!

## Reduce file watcher memory usage
Previously, Metals consumed a huge amount of memory for file watchers in large projects, especially on macOS.

> Now, Metals uses a memory-efficient way to watch files to detect changes and consumes less memory.

Reviewer 2: ```suggestion
Now, Metals uses a memory-efficient way to watch files to detect changes and consume less memory.
```

For more technical details, see the original PR: [\#3758](https://github.com/scalameta/metals/pull/3758).

## [Scala3] Override Completions

> Override completions for Scala3 is available with Metals 0.11.6!

Reviewer 2: ```suggestion
Override completions for Scala3 are now available with Metals 0.11.6!
```

Now, it's possible to implement methods from the super class quite easily in Scala3.

![override-completion](https://i.imgur.com/Go3sMxy.gif)

## [Scala3] Show scaladoc on hover for Scala 3 project

> Now, Metals shows the scaladoc on hover for Scala3 project. (Before this release, Metals unable to show the scaladoc for the symbols from third-party modules).

Reviewer 2: ```suggestion
Now, Metals shows the scaladoc on hover for Scala3 projects. (Before this release, Metals was unable to show the scaladoc for the symbols from third-party modules).
```

![hover-scala3](https://i.imgur.com/Svzq5DD.png)

### [Scala3] Support `completionItem/resolve`

> Now, in Scala 3 we get scaladoc and shows default values when moving accross completions. 

Reviewer 2: ```suggestion
Now, in Scala 3 we get scaladoc that shows the default values when moving across completions. 
```

![completion-item-resolve](https://i.imgur.com/Tz6AOsx.gif)

## Show parent scaladoc if implementation is returning empty

> Scaladocs can be inspected whenever you hover, use a completion or signature help. Up until recently we only showed you the documentation if the exact method you are using had the scaladocs written, which meant that if you overrode a method and didn't add the scaladoc comments again we would not show you any documentation. One example of such method could `headOption` on a `List`.

Reviewer 2: ```suggestion
Scaladocs can be inspected whenever you hover, use a completion or signature help. Up until recently we only showed you the documentation if the exact method you are using had the scaladocs written, which meant that if you overrode a method and didn't add the scaladoc comments again we would not show you any documentation. One example of such method is `headOption` on `List`.
```

 From this release we will also search the parent method in case the current method's scaladoc are empty. 

## [Scala 3] Show scaladocs for signature help

As mentioned in the previous paragraph, Metals can show you documentation in three different places. That, however, was true only for Scala 2 previously. In this release, we will now show you proper documentation whenever invoking signature help.

> As a remainder, signature help is used to indicate what parameters can be used in a method. It should pop up automatically after writing `(`, but you can also invoke it manually. In VS Code that takes the form of `editor.action.triggerParameterHints` command, which can also be bound to a shortcut and by default is.

Reviewer 2: ```suggestion
As a reminder, signature help is used to indicate what parameters can be used in a method. It should pop up automatically after writing `(`, but you can also invoke it manually. In VS Code that takes the form of `editor.action.triggerParameterHints` command, which can also be bound to a shortcut and by default is.
```

## [MUnit] Test Explorer can find helper methods from parent classes

> MUnit allows to use [helper functions](https://scalameta.org/munit/docs/tests.html#declare-tests-inside-a-helper-function) when declaring  tests. Very often those helper methods are extracted to some parent classes which are extended by many test suites.  

Reviewer 2: ```suggestion
MUnit allows to use [helper functions](https://scalameta.org/munit/docs/tests.html#declare-tests-inside-a-helper-function) when declaring tests. Very often those helper methods are extracted to some parent classes which are extended by many test suites.  
```

Now, Metals can find usages of those helper methods and display them in Test Explorer.  

> This feature is available for Bloop and SBT 1.7.0-M2 or later.

Author: I think we can mention sbt 1.7.0-M2, as https://github.com/sbt/sbt/issues/6896 is a separate problem with this feature :) (anyway, I'll try to fix the problem in sbt).

> This feature is available for Bloop and SBT 1.7.0-M2 or later.

Reviewer: I'm afraid that some people will try `sbt 1.7.0-M2` and then come across https://github.com/sbt/sbt/issues/6896, and this could be very misleading experience that "Metals are not working". 
Alternatively, both `sbt 1.7.0-M2` and https://github.com/sbt/sbt/issues/6896 can be mentioned and then users will have a full picture.

> This feature is available for Bloop and SBT 1.7.0-M2 or later.

Author: Ok then, we can drop sbt 1.7.0-M2 from release note :+1: (I don't have strong opinion).

![MUnit-helper-methods](https://imgur.com/GGRDpXA.gif)

## Support Cats Effect stacktraces in stacktrace analyzer

Cats Effect offers [asynchronous stack tracing](https://typelevel.org/cats-effect/docs/tracing#asynchronous-stack-tracing) which augments exceptions with additional information.  

> Now, Stacktrace Analyzer is able to recognize CE's stacktraces and provide link to location in code. Say no to tedious debugging when you have only stacktrace from logs!

Reviewer 2: ```suggestion
Now, Stacktrace Analyzer is able to recognize CE's stacktraces and provide link to location in code. Say no to tedious debugging when you have only stacktrace from the logs!
```

![cats-effect-stacktraces](https://imgur.com/5fMvcYd.gif)

## Improve implement all completion and code action

> Previously, when we invoke implement all completion and code action, Metals fills with something like `x$0` for Java parameter names. Now, Metals fills the correct parameter names for Java parameters.

Reviewer 2: ```suggestion
Previously, when we invoked the "implement all members" completion and code action, Metals used `x$0` for Java parameter names. Now, Metals fills these with the correct parameter names.
```
